### DISCUSSION TITLE
motr interface to Jupyter notebook

### DISCUSSION THREAD

**Discussion by yanqingfu** [Seagate/cortx#464] [Show and tell] [answered]
1.  [Install Jupyter Notebook](https://www.youtube.com/watch?v=D_rkUgFp0co&t=1s) 

```
yum -y install gcc gcc-c++ kernel-devel install python-devel libxslt-devel libffi-devel openssl-devel python-pip

pip install --upgrade pip

python --version && pip --version

pip install jupyter --use-feature=2020-resolver
```

   enable Python 3 kernel for Jupter Notebook if needed

```
yum install -y python3

python3 -m pip install ipykernel

python3 -m ipykernel install --user
```

2. Start Jupyter Notebook in root 

`jupyter notebook --no-browser --allow-root`

![image](https://user-images.githubusercontent.com/68423216/98118048-8e610880-1ee5-11eb-82c3-23b53203fcad.png)

3. Config as [Running a Jupyter notebook from a remote server](https://ljvmiranda921.github.io/notebook/2018/01/31/running-a-jupyter-notebook/) 

when open localhost, enter the token shows in step 2
![token](https://user-images.githubusercontent.com/68423216/98118086-a20c6f00-1ee5-11eb-89b5-87ff91b561fc.JPG)

you can run Python and terminal at the same time

![image](https://user-images.githubusercontent.com/68423216/98122401-82784500-1eeb-11eb-8f74-56dcf988ede2.png)

4. [run Jupyter notebook command line](https://jakevdp.github.io/PythonDataScienceHandbook/01.05-ipython-and-shell-commands.html)

Environment: RHEL7.7  with 3.10.0-1062.el7 kernel, cloud Linux VM with installed cortx-motr package

Jupyter output: 
![image](https://user-images.githubusercontent.com/68423216/98095161-2995b600-1ec5-11eb-8313-b1e00bb9e9f8.png)

![image](https://user-images.githubusercontent.com/68423216/98095239-47fbb180-1ec5-11eb-8b1f-a33dfe17529f.png)

more details: [jupytermotrtest.pdf](https://github.com/Seagate/cortx/files/5486902/jupytermotrtest.pdf)

This opens two main opportunities:

1. public cloud VM for testing purpose

for testing purpose, we can set up a "public" cloud VM with CORTX installed and Jupyter notebook running

for people who want to test the CORTX, we can provide the Jupyter token when people request or just provide the token

In this way, we can make testing easier.  People just need to use ssh command to connect the cloud VM while using a local PC without installing anything.

2. install packages by using the Jupyter terminal

![image](https://user-images.githubusercontent.com/68423216/98116615-7f795680-1ee3-11eb-907d-7ea650f32ba0.png)

![image](https://user-images.githubusercontent.com/68423216/98116664-8e600900-1ee3-11eb-8c20-756ed274b974.png)

![image](https://user-images.githubusercontent.com/68423216/98118224-ded86600-1ee5-11eb-820f-7e127547ffb0.png)

Also tried the following:

**SSH forwarding method**

![image](https://user-images.githubusercontent.com/68423216/98095463-9741e200-1ec5-11eb-86ae-28922f2234a3.png)

  1. installed google-chrome then run `jupyter notebook` seems google-chrome has bugs
      
![image](https://user-images.githubusercontent.com/68423216/98096324-965d8000-1ec6-11eb-9330-846b0cf6c9fd.png)

  2. installed firefox, then run `./firefox/firefox`, no error, but the rendering is very slow

![image](https://user-images.githubusercontent.com/68423216/98096520-d15fb380-1ec6-11eb-8a2f-f0d2463b53a5.png)

**Comment by johnbent** [comment]
Wow, @yanqingfu , this is awesome!  I'm not very familiar with Jupyter notebook though.  What exactly is this doing?  Would this allow someone to run S3 clients in a Jupyter notebook and connect to an instance of CORTX running elsewhere?  Or run motr clients in a Jupyter notebook and connect to an instance of CORTX running elsewhere?  Is this allowing running CORTX server in a Jupyter notebook?!?!

**Comment by yanqingfu** [comment, marked as answer]
@johnbent Hi John, you can think of Jupyter Notebook as a (mirrored) WEB terminal here. If you installed Jupyter Notebook on the remote cloud VM, you can access all your remote machine content through SSH (need to know remote host user info, remote host address, and maybe root access(password)) on your local PC's Jupyter Notebook.

-remote machine without cortx package installed, I can use my local Jupyter Notebook terminal to install and test cortx package
-remote machine with cortx package installed, I can use my local Jupyter Notebook Ipython environment to test cortx package

Basically, with your local Juypter Notebook, you can run all the [instructions ](https://github.com/Seagate/cortx-motr/blob/main/doc/Quick-Start-Guide.rst), plus you can do result visualizations, etc.

> Would this allow someone to run S3 clients in a Jupyter notebook and connect to an instance of CORTX running elsewhere? Or run motr clients in a Jupyter notebook and connect to an instance of CORTX running elsewhere? Is this allowing running CORTX server in a Jupyter notebook?!?!

**If the current config can do all of those, I believe Jupyter Notebook can do too.
Better to have some code example, so that I can do a double-check.**
